File: baekjoon/해시/7785.cpp
#include <iostream>
#include <string>
#include <map>
#include <algorithm>

using namespace std;

bool compare(pair<string, bool> a, pair<string, bool> b) {
    return a.first > b.first;
}

int main() {
    int n;
    cin >> n;

    unordered_map<string, bool> logs;
    while (n--) {
        string user, status;
        cin >> user >> status;
        if (status == "enter") {
            logs[user] = true;
        } else {
            logs[user] = false;
        }
    }

    vector<pair<string, bool>> exist;
    unordered_map<string, bool>::iterator iter = logs.begin();
    while (iter != logs.end()) {
        if (iter->second) exist.push_back(*iter);
        iter++;
    }

    sort(exist.begin(), exist.end(), compare);

    for (pair<string, bool> info : exist) {
        cout << info.first << '\n';
    }

    return 0;
}

File: baekjoon/해시/7785.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st;

        Map<String, Boolean> logs = new HashMap<>();

        int n = Integer.parseInt(br.readLine());
        for (int i = 0; i < n; ++i) {
            st = new StringTokenizer(br.readLine());
            String name = st.nextToken();
            String status = st.nextToken();

            if (status.equals("enter")) {
                logs.put(name, true);
            } else {
                logs.put(name, false);
            }
        }

        List<String> exist = new ArrayList<>();
        for (String s : logs.keySet()) {
            if (logs.get(s)) {
                exist.add(s);
            }
        }

        exist.sort(new CustomComparator());

        for (String s : exist) {
            bw.write(s);
            bw.write('\n');
        }
        bw.flush();
        bw.close();
    }
}

class CustomComparator implements Comparator<String> {
    @Override
    public int compare(String a, String b) {
        int compareValue = a.compareTo(b);
        if (compareValue > 0) return -1;
        if (compareValue == 0) return 0;
        return 1;
    }
}